src/core/lib/iomgr/event_engine_shims/endpoint.cc

```diff
@@ -0,0 +1,341 @@
+// Copyright 2022 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include <grpc/support/port_platform.h>
+
+#include "src/core/lib/iomgr/event_engine_shims/endpoint.h"
+
+#include <atomic>
+#include <memory>
+
+#include "absl/status/status.h"
+#include "absl/strings/string_view.h"
+
+#include <grpc/event_engine/event_engine.h>
+#include <grpc/impl/codegen/slice.h>
+#include <grpc/slice.h>
+#include <grpc/slice_buffer.h>
+#include <grpc/support/log.h>
+#include <grpc/support/time.h>
+
+#include "src/core/lib/event_engine/posix.h"
+#include "src/core/lib/event_engine/tcp_socket_utils.h"
+#include "src/core/lib/gprpp/sync.h"
+#include "src/core/lib/iomgr/closure.h"
+#include "src/core/lib/iomgr/endpoint.h"
+#include "src/core/lib/iomgr/error.h"
+#include "src/core/lib/iomgr/port.h"
+#include "src/core/lib/transport/error_utils.h"
+
+extern grpc_core::TraceFlag grpc_tcp_trace;
+
+namespace grpc_event_engine {
+namespace experimental {
+
+namespace {
+
+constexpr int64_t kShutdownBit = static_cast<int64_t>(1) << 32;
+
+// A wrapper class to manage Event Engine endpoint ref counting and
+// asynchronous shutdown.
+class EventEngineEndpointWrapper {
+ public:
+  struct grpc_event_engine_endpoint {
+    grpc_endpoint base;
+    EventEngineEndpointWrapper* wrapper;
+    std::aligned_storage<sizeof(SliceBuffer), alignof(SliceBuffer)>::type
+        read_buffer;
+    std::aligned_storage<sizeof(SliceBuffer), alignof(SliceBuffer)>::type
+        write_buffer;
+  };
+
+  explicit EventEngineEndpointWrapper(
+      std::unique_ptr<EventEngine::Endpoint> endpoint);
+
+  int Fd() {
+    grpc_core::MutexLock lock(&mu_);
+    return fd_;
+  }
+
+  absl::string_view PeerAddress() {
+    grpc_core::MutexLock lock(&mu_);
+    return peer_address_;
+  }
+
+  absl::string_view LocalAddress() {
+    grpc_core::MutexLock lock(&mu_);
+    return local_address_;
+  }
+
+  void Ref() { refs_.fetch_add(1, std::memory_order_relaxed); }
+  void Unref() {
+    if (refs_.fetch_sub(1, std::memory_order_acq_rel) == 1) {
+      delete this;
+    }
+  }
+
+  // Returns a managed grpc_endpoint object. It retains ownership of the
+  // object.
+  grpc_endpoint* GetGrpcEndpoint() { return &eeep_->base; }
+
+  // Read using the underlying EventEngine endpoint object.
+  inline void Read(absl::AnyInvocable<void(absl::Status)> on_read,
```

Comment:
Is `inline` terribly meaningful here? This is a cc file, this code should only show up in a single translation unit right?

Reply:
Removed.